### PR TITLE
NZSL-51: Refactor Freelex code into a module, add bridge class to resolve correct class

### PR DESCRIPTION
**Pull request by joshmcarthur**
This pull request is the first step on NZSL-51, moving code that relies on data from Freelex into a module. This includes the signs controller and related services (primarily related to search).

This PR also adds a model named `SignModel`, with a class method to `resolve` the correct model class. This class is used outside the `Freelex` module namespace to resolve the correct model class to use - e.g. when querying for random signs, sign of the day, or for looking up signs by ID. At the moment, this class always returns `Freelex::Sign`. The next PR will start adding support for `Signbank::Sign`. 

The objective of this change is to enable support for feature flagging the source for dictionary data. This allows us to switch between Freelex and Signbank while testing, to ensure data equivalency and to be able to test dictionary export changes. This change is going to stay in place until at least February, at which point we'll be able to remove the Freelex-based code, and replace `SignModel.resolve` with direct calls to `Signbank::Sign`.

### CODE REVIEW

File: app/controllers/freelex/signs_controller.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      end
+
+      render json: AutocompleteSearchService.new(search_term: search_term_param).find_suggestions
+    end
+
+    private
+
+    ##
+    # Continue to render templates from app/views/signs
+    def default_render
```
Comment (mischa-s) [inline app/controllers/freelex/signs_controller.rb:45]: TIL 🔍

File: app/models/freelex/sign.rb [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+      # Fetch a sign for the given ID. The sign is fetched from `Rails.cache` if
+      # possible, otherwise it is fetched from Freelex and then cached for later
+      # use.
+      # @param sign_id [String]
+      # @return [Sign] if we could find a sign
+      # @return [nil] if we could not find a sign
+      def fetch_by_id_via_cache(sign_id)
+        return first(id: sign_id) unless FeatureFlags::StoreVocabSheetItemsInRailsCache.enabled?
+
+        # We use the more verbose `Rails.cache.fetch` and `Rails.cache.write`
```
Comment (mischa-s) [inline app/models/freelex/sign.rb]:
```suggestion
        # We use the more verbose `Rails.cache.read` and `Rails.cache.write`
```

File: app/models/freelex/sign.rb [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+      # @return [Sign] if we successfully found a sign
+      # @return [nil] if we did not find a sign
+      def random
+        first(random: 1)
+      end
+
+      # @return [Sign]
+      def sign_of_the_day
+        # We use the more verbose `Rails.cache.fetch` and `Rails.cache.write`
```
Comment (mischa-s) [inline app/models/freelex/sign.rb]:
```suggestion
        # We use the more verbose `Rails.cache.read` and `Rails.cache.write`
```

File: app/models/freelex/sign_menu.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      ].freeze,
+      [
+        ['7.1.1', '7.1.2', '7.1.3', '7.1.4'].freeze,
+        ['7.2.1'].freeze,
+        ['7.3.1', '7.3.2', '7.3.3'].freeze,
+        ['7.4.1', '7.4.2'].freeze
+      ].freeze
+    ].freeze
+
+    LOCACTIONS = [
```
Comment (mischa-s) [inline app/models/freelex/sign_menu.rb]:
```suggestion
    LOCATIONS = [
```

File: app/models/freelex/sign_menu.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      ['Weather',                                   41].freeze,
+      ['Work',                                      42].freeze
+    ].freeze
+
+    def self.handshapes
+      HANDSHAPES
+    end
+
+    def self.locations
+      LOCACTIONS
```
Comment (mischa-s) [inline app/models/freelex/sign_menu.rb]:
```suggestion
      LOCATIONS = [
```

File: app/models/freelex/sign_parser.rb [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    def parse_booleans
+      BOOLEAN_MAP.each do |key, tag|
+        value = @data.value_for_tag(tag).to_bool
+        @sign.send("#{key}=", value)
+      end
+    end
+
+    def parse_examples
+      # examples
```
Comment (mischa-s) [inline app/models/freelex/sign_parser.rb]:
```suggestion
```

**Review by mischa-s — COMMENTED** [review]
Found a few typos but other than that seems fine. I dont have much context on the code, but I don't see any obvious logic issues or errors

**Comment by Br3nda** [comment]
This is so much better than the Sign model code. Great change.

**Review by mischa-s — APPROVED** [review]

**Pull request merged**